### PR TITLE
PERF-161: Hashing improvements

### PR DESCRIPTION
**Pull request by snaumenko-st**
Improved memory usage by using Spans and shared array pool

### CODE REVIEW

**Review by SergeiPavlov — APPROVED** [review]

File: Extensions/Xtensive.Orm.Security/Cryptography/GenericHashingService.cs [hunk 1]
```diff
@@ -28,70 +29,72 @@ public abstract class GenericHashingService : IHashingService
...
+    private void ComputeHashInternal(string password, ReadOnlySpan<byte> salt, Span<byte> hash)
+    {
+      var encoding = Encoding.UTF8;
+      var buffer = ArrayPool<byte>.Shared.Rent(salt.Length + encoding.GetMaxByteCount(password.Length));
+      try {
+        salt.CopyTo(buffer);
+        var written = encoding.GetBytes(password.AsSpan(), buffer.AsSpan(salt.Length));
+
+        using var hasher = GetHashAlgorithm();
+        _ = hasher.TryComputeHash(new ReadOnlySpan<byte>(buffer, 0, salt.Length + written), hash, out _);
```
Comment (SergeiPavlov) [inline Extensions/Xtensive.Orm.Security/Cryptography/GenericHashingService.cs]: It would be good to check return value & `bytesWritten == HashSizeInBytes`

Comment (snaumenko-st) [reply Extensions/Xtensive.Orm.Security/Cryptography/GenericHashingService.cs]: It's not necessary at runtime, so I added `Debug` asserts for that.

Comment (SergeiPavlov) [reply Extensions/Xtensive.Orm.Security/Cryptography/GenericHashingService.cs]: `Debug.Assert` is almost useless. Nobody runs DEBUG build, except for local unit testing.
Why not to check and throw an exception? It is lightweight operation comparing  Crypto-hashing.

The rule should be: if some API returns "Success status"  - check it always.

Comment (snaumenko-st) [reply Extensions/Xtensive.Orm.Security/Cryptography/GenericHashingService.cs]: Yes, it's useless in runtime, of course. But the tests won't pass if there is an issue somewhere.
This call returns `false` only if buffer is not long enough. But at the same time this buffer is created by this class and can't be too short except if there is some bug introduced later. The same statement is true for other checks like `hash.Length`, `salt.Length`, etc. It's a common practice in `dotnet runtime` repo also. We just have to support good enough test coverage to not break it.

Comment (snaumenko-st) [reply Extensions/Xtensive.Orm.Security/Cryptography/GenericHashingService.cs]: Ok, fixed.

**Pull request merged**